mesonbuild/coredata.py

```diff
@@ -103,8 +103,11 @@ def printable_value(self) -> T.Union[str, int, bool, T.List[T.Union[str, int, bo
     def validate_value(self, value: T.Any) -> _T:
         raise RuntimeError('Derived option class did not override validate_value.')
 
-    def set_value(self, newvalue: T.Any) -> None:
-        self.value = self.validate_value(newvalue)
+    def set_value(self, newvalue: T.Any) -> bool:
+        newvalue = self.validate_value(newvalue)
+        dirty = not hasattr(self, "value") or self.value != newvalue  # pylint: disable=access-member-before-definition
+        self.value = newvalue
+        return dirty
```

Comment:
mypy does not seems to like your way, but it accepts like that:

```suggestion
    def set_value(self, newvalue: T.Any) -> bool:
        oldvalue = getattr(self, 'value', None)
        self.value = self.validate_value(newvalue)
        return self.value != oldvalue
```